pkg/master/controller.go

```diff
@@ -284,146 +285,154 @@ type EndpointReconciler interface {
 	// managed only by ReconcileEndpoints; therefore, to understand this, you need only
 	// understand the requirements.
 	//
+	// It uses auxilary config map to keep expiration times for apiservers.
+	// Each apiserver separately update its expiration time and removes all expired apiservers.
+	//
 	// Requirements:
 	//  * All apiservers MUST use the same ports for their {rw, ro} services.
 	//  * All apiservers MUST use ReconcileEndpoints and only ReconcileEndpoints to manage the
 	//      endpoints for their {rw, ro} services.
 	//  * ReconcileEndpoints is called periodically from all apiservers.
-	ReconcileEndpoints(serviceName string, ip net.IP, endpointPorts []api.EndpointPort, reconcilePorts bool) error
+	ReconcileEndpoints(configmapName, serviceName string, ip net.IP, endpointPorts []api.EndpointPort, reconcilePorts bool, now time.Time) error
 }
 
-// masterCountEndpointReconciler reconciles endpoints based on a specified expected number of
-// masters. masterCountEndpointReconciler implements EndpointReconciler.
-type masterCountEndpointReconciler struct {
-	masterCount    int
-	endpointClient coreclient.EndpointsGetter
+// dynamicEndpointReconciler reconciles endpoints based on a config map with expiration times.
+// dynamicEndpointReconciler implements EndpointReconciler.
+type dynamicEndpointReconciler struct {
+	endpointClient  coreclient.EndpointsGetter
+	configmapClient coreclient.ConfigMapsGetter
 }
 
-var _ EndpointReconciler = &masterCountEndpointReconciler{}
+var _ EndpointReconciler = &dynamicEndpointReconciler{}
+
+// NewDynamicEndpointReconciler creates a new EndpointReconciler that reconciles based on an auxiliary config map.
+func NewDynamicEndpointReconciler(endpointClient coreclient.EndpointsGetter,
+	configmapClient coreclient.ConfigMapsGetter) *dynamicEndpointReconciler {
+	return &dynamicEndpointReconciler{
+		endpointClient:  endpointClient,
+		configmapClient: configmapClient,
+	}
+}
 
-// NewMasterCountEndpointReconciler creates a new EndpointReconciler that reconciles based on a
-// specified expected number of masters.
-func NewMasterCountEndpointReconciler(masterCount int, endpointClient coreclient.EndpointsGetter) *masterCountEndpointReconciler {
-	return &masterCountEndpointReconciler{
-		masterCount:    masterCount,
-		endpointClient: endpointClient,
+// updateCM updates TTL configmap for API servers: renews TTL for key server and removes all expired entries.
+func updateCM(key string, m *api.ConfigMap, now time.Time) {
+	const format = "20060102 150405 MST"
+	window := 3 * DefaultEndpointReconcilerInterval
+	newData := map[string]string{}
+	for k, v := range m.Data {
+		ttl, err := time.Parse(format, v)
+		if err != nil {
+			continue
+		}
+		if now.Before(ttl) {
```

Comment:
This would be the first thing we add that doesn't converge in the presence of time drift, so I don't think it's acceptable to say "it doesn't handle time drifts".  We do assume time synchronization but I don't think saying that HA of the masters depends on time sync is the logical next step.  

If there's any place in Kube where time drift is critically relevant, it's here.